### PR TITLE
recover from inconsistent database state (close #231)

### LINKED ISSUE #231: Recover from an inconsistent database state
@0x777 suggested that graphql-engine will ignore the errors ( Due to inconsistent db state ) if it has occured due to updates to db via psql or any client. Graphql engine will provide an endpoint which will be called by console to show instructions to the user on how to fix it.

@praveenweb

### PR DESCRIPTION
**Pull request by rakeshkky**
### Description

The server keeps track of all inconsistent metadata objects and reports them on `/v1/query` of type `get_inconsistent_metadata`.

API request:-
```http
POST /v1/query
{
  "type": "get_inconsistent_metadata",
  "args": {}
}
```
Response:-
```json
{ 
    "is_consistent": false,
    "inconsistent_objects": [
  {
    "definition": {
      "role": "user",
      "comment": null,
      "permission": {
        "allow_aggregations": false,
        "columns": [
          "author",
          "author_id",
          "id",
          "title"
        ],
        "filter": {
          "author": {
            "_eq": "X-Hasur\r\na-User-Id"
          }
        }
      },
      "table": "article"
    },
    "reason": "table \"article\" does not exist",
    "type": "select_permission"
  },
  {
    "definition": {
      "using": {
        "foreign_key_constrai\r\nnt_on": "author_id"
      },
      "name": "author_obj",
      "comment": null,
      "table": "article"
    },
    "reason": "table \"article\" does not exist",
    "type": "object_relation"
  },
  {
    "definition": {
      "using": {
        "foreign_key_constraint_on": {
          "column": "author_id",
          "table": "article"
        }
      },
      "name": "articles",
      "comment": null,
      "table": "author"
    },
    "reason": "t\r\nable \"article\" does not exist",
    "type": "array_relation"
  },
  {
    "definition": "article",
    "reason": "no such table/view exists in postgres : \"article\"",
    "type": "table"
  }
]
}
```
Use `drop_inconsistent_metadata` query type to purge all inconsistent objects. 

API
```http
POST /v1/query
{
  "type": "drop_inconsistent_metadata",
  "args": {}
}
```
Response:-
```json
{
    "message": "success"
}
```

### Affected components 

- Server
- Console
- Docs
- Tests

### Related Issues

close #231 

### Solution and Design

1. `buildSchemaCache` function from `Hasura.RQL.DDL.Schema.Table` module is updated to collect inconsistent objects and push to schema cache.
2. Add `get_inconsistent_metadata` and `drop_inconsistent_metadata` query types in `/v1/query` API
3. Remote schemas related metadata API code is refactored to support handling inconsistent ones. 

### Steps to test and verify

1. Start GraphQL Engine, add few tables and define permissions and relationships.
2. Through `psql` or `adminer` drop any table which is being tracked.
3. Now, restart GraphQL Engine or `reload metadata` in settings section of console. 
4. Make `get_inconsistent_metadata` API query (spec given above), you'll get list of all inconsistent metadata objects.

### CODE REVIEW

**Comment by danielcompton** [comment]
Thanks for this!

We were able to use the soft start feature to boot an app when our schema had changed externally. Previous Hasura wouldn't boot, so we couldn't reset or apply new metadata. This looks to be on the right track. I didn't try the API though, just using the CLI and the console.

**Comment by rakeshkky** [comment]
@wawhal 

An example inconsistent object for remote schema
```json
[
    {
        "definition": {
            "definition": {
                "url": "http://localhost:8181/v1alpha1/graphql",
                "headers": [],
                "url_from_env": null,
                "forward_client_headers": false
            },
            "name": "chinook",
            "comment": null
        },
        "reason": "HTTP exception occurred while sending the request to http://localhost:8181/v1alpha1/graphql",
        "type": "remote_schema"
    }
]
```

File: server/tests-py/test_inconsistent_meta.py [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+            q={
+                "type": "replace_metadata",
+                "args": export
+            }
+        )
+        assert st_code == 400, resp
+
+        # drop inconsistent objects
+        st_code, resp = hge_ctx.v1q(q=self.drop_inconsistent_metadata)
+        assert st_code == 200, resp
```
Comment (0x777) [inline server/tests-py/test_inconsistent_meta.py]: fetch inconsistent objects after this

File: server/src-lib/Hasura/Server/Init.hs [hunk 2]
```diff
@@ -782,3 +783,9 @@ serveOptsToLog so =
 mkGenericStrLog :: T.Text -> String -> StartupLog
 mkGenericStrLog k msg =
   StartupLog L.LevelInfo k $ J.toJSON msg
+
+inconsistentObjsLog :: SchemaCache -> StartupLog
+inconsistentObjsLog sc =
+  StartupLog L.LevelInfo "inconsistent_objects" infoVal
```
Comment (0x777) [inline server/src-lib/Hasura/Server/Init.hs]: inconsistent_metadata

File: server/src-lib/Hasura/Server/App.hs [hunk 3]
```diff
@@ -236,22 +239,18 @@ v1QueryHandler query = do
...
     dbActionReload = do
       (resp, newSc) <- dbAction
       httpMgr <- scManager . hcServerCtx <$> ask
       --FIXME: should we be fetching the remote schema again? if not how do we get the remote schema?
-      newGCtxMap <- GS.mkGCtxMap (scTables newSc) (scFunctions newSc)
-      (mergedGCtxMap, defGCtx) <-
-        mergeSchemas (scRemoteResolvers newSc) newGCtxMap httpMgr
-      let newSc' =
-            newSc { scGCtxMap = mergedGCtxMap, scDefaultRemoteGCtx = defGCtx }
+      newSc' <- flip resolveRemoteSchemas httpMgr =<< GS.updateSCWithGCtx newSc
```
Comment (0x777) [inline server/src-lib/Hasura/Server/App.hs]: use `>>=` instead of `=<<`

File: server/src-lib/Hasura/RQL/Types/SchemaCacheTypes.hs [hunk 4]
```diff
@@ -67,3 +70,18 @@ data SchemaDependency
 
 $(deriveToJSON (aesonDrop 2 snakeCase) ''SchemaDependency)
 instance Hashable SchemaDependency
+
+data InconsistentSchemaObj
```
Comment (0x777) [inline server/src-lib/Hasura/RQL/Types/SchemaCacheTypes.hs]: inconsistent metadata object

File: server/src-lib/Hasura/RQL/Types/SchemaCacheTypes.hs [hunk 5]
```diff
@@ -67,3 +70,18 @@ data SchemaDependency
...
+  = InconsistentSchemaObj
+  { _soId     :: !SchemaObjId
+  , _soType   :: !T.Text
```
Comment (0x777) [inline server/src-lib/Hasura/RQL/Types/SchemaCacheTypes.hs]: use a sum type instead of text

Comment (0x777) [reply server/src-lib/Hasura/RQL/Types/SchemaCacheTypes.hs]: `MetadataObjectType`

File: server/src-lib/Hasura/RQL/DDL/Schema/Table.hs [hunk 6]
```diff
@@ -107,30 +107,40 @@ runTrackTableQ q = do
   trackExistingTableOrViewP2 (tName q) False
 
 purgeDep :: (CacheRWM m, MonadTx m)
```
Comment (0x777) [inline server/src-lib/Hasura/RQL/DDL/Schema/Table.hs]: define `purgeMetadataObj`

File: server/src-lib/Hasura/RQL/Types/SchemaCacheTypes.hs [hunk 7]
```diff
@@ -28,13 +29,16 @@ data SchemaObjId
   | SOQTemplate !TQueryName
   | SOTableObj !QualifiedTable !TableObjId
   | SOFunction !QualifiedFunction
+  | SORemoteSchema !RemoteSchemaName
```
Comment (0x777) [inline server/src-lib/Hasura/RQL/Types/SchemaCacheTypes.hs]: Create new types for `MetadataObj`, `MetadataObjId`, `MetadataObjType`.

**Comment by rikinsk** [comment]
@dsandip @wawhal  This will make https://docs.hasura.io/1.0/graphql/manual/troubleshooting/index.html redundant right?

**Comment by rikinsk** [comment]
related #415

**Comment by wawhal** [comment]
@rikinsk I think it should be there. Because finding those inconsistencies in the schema is a not straight forward. We should just add a note at the top saying "If your version is above <version>, you can look at all the inconsistencies at `/metadata/status`"

**Pull request merged**